File: Problem1.java
// ## Problem1 (https://leetcode.com/problems/binary-search-tree-iterator/)
// Time Complexity : Ammortized O(1)
// Space Complexity : O(H), where H is the height of input tree
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


class BSTIterator {
    // Simulate an inOrder traversal call-stack
    private Stack<TreeNode> inOrderSt;

    private void helper(TreeNode root){
        if(root==null){
            return;
        }
        while(root!=null){
            this.inOrderSt.add(root);
            root=root.left;
        }
    }

    public BSTIterator(TreeNode root) {
        inOrderSt=new Stack<>();
        helper(root);
    }
    
    public int next() {
        TreeNode curr=inOrderSt.pop();
        // Add right sub-tree to stack
        helper(curr.right);
        return curr.val;
    }
    
    public boolean hasNext() {
        return inOrderSt.size()!=0;
    }
}

File: Problem1.js
// Time Complexity : next - Worst Case -O(h), Avg O(1), hasNext O(1)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

var stack;
var BSTIterator = function (root) {
    stack = [];
    dfs(root)
};

/**
 * @return {number}
 */
BSTIterator.prototype.next = function () {
    let result = stack.pop();
    dfs(result.right);
    return result.val;
};

/**
 * @return {boolean}
 */
BSTIterator.prototype.hasNext = function () {
    return (!(stack.length === 0));
};

var dfs = function (root) {
    while (root) {
        stack.push(root);
        root = root.left;
    }
}

